### ISSUE TITLE
ValueGeneratorCache holds scoped ValueGeneratorSelector references

### ISSUE THREAD

**Issue by FStolte** [dotnet/efcore#24637]
_**tl;dr:** The `ValueGeneratorSelector` captures a reference to itself in the factory closure that is passed to `ValueGeneratorCache.GetOrAdd()`, since the closure contains the non-static method `Create()`. Therefore, the `ValueGeneratorSelector` will not be garbage collected once its scope is disposed and remains in memory indefinitely despite being a scoped service._

So I've been hunting a memory leak in our Blazor-App for a while now, and after finding the main culprit, I saw some remaining DbContexts in memory that I couldn't make sense of. Although they weren't hogging that much memory, I decided to continue my investigation. From the GC dump I could only see types of EF-Core and Npgsql in the chain holding on to those contexts, so I started digging through the code of those two projects.
After looking for a while I think I have found the reason for this behavior, and if I'm right (and I'm confident I am, but feel free to correct me if I'm not), it is due to an unfortunate interaction between the EF-Core `ValueGeneratorSelector` and it's Npgsql sub-class `NpgsqlValueGeneratorSelector`.

Here's the `Select()`-method from the `ValueGeneratorCache`:
```
public virtual ValueGenerator Select(IProperty property, IEntityType entityType)
{
    Check.NotNull(property, nameof(property));
    Check.NotNull(entityType, nameof(entityType));

    return Cache.GetOrAdd(property, entityType, (p, t) => CreateFromFactory(p, t) ?? Create(p, t));
}
```

First of all, the documentation for `ValueGeneratorCache` says "The service lifetime is ServiceLifetime.Singleton", while `ValueGeneratorSelector` says "The service lifetime is ServiceLifetime.Scoped". Therefore, `ValueGeneratorCache` must not use `ValueGeneratorSelector`.
As you can see, the closure that's passed as a factory function includes a call to the `Create()` method of `ValueGeneratorSelector`, which is **not static**. Therefore, the closure includes a reference to the `ValueGeneratorSelector` instance. When `ValueGeneratorCache` caches this factory function, it holds on to the `ValueGeneratorSelector` instance as well, thereby preventing it from being garbage collected, even after it's scope has been disposed.
Now, the `NpgsqlValueGeneratorSelector` holds, among other things, a reference to the `NpgsqlRelationalConnection`, which in turn holds a reference to the DbContext for which it was created. Therefore, the DbContext and its tracked entities are also prevented from garbage collection.

If the documentation about the lifetime of these EF-Core classes is up to date, I think that the fix for this issue also lies within EF-Core. In fact, extracting the contents of `Create()` into a static method and calling that one from `Select()`  should be sufficient to fix this without introducing a breaking change, if I am not overlooking anything.

In case it helps, here is the relevant excerpt from the heap that I analyzed, the root node in this case is the `ServiceProviderCache`
![Heap](https://user-images.githubusercontent.com/18054693/114414343-03c8b780-9baf-11eb-96d3-da0bc0a4ea0a.png)

I'm looking forward to hearing whether this is an actual bug, or if I just overlooked something in my frenzy to find an explanation for our leak 😁

EF Core version: 5.0.2
Database provider: Npgsql.EntityFrameworkCore.PostgreSQL, Version: 5.0.2
Target framework: .NET 5.0
Operating system: Docker container based on `mcr.microsoft.com/dotnet/aspnet:5.0`
IDE: Visual Studio Enterprise 2019 16.9.1

**Comment by FStolte** [comment]
Seems my idea of a non-breaking fix was a bit too naive, since there can be a derived `ValueGeneratorSelector` that overrides `Create()` and expects it to be called in `Select()`...

**Comment by snappyfoo** [comment]
Just commenting as someone interested in this find: to me it’s odd that the factory forms part of the cache key as that seems to be the cause of the problem. Looking at the history it hasn’t always used the factory in the key but I do not understand all the reasons why it might have been changed to do so.

**Comment by ajcvickers** [comment]
Note for triage: I think the way to fix this is to move away from the selector, which was born out of one-model-to-rule-them-all, and instead configure the value generator type directly in the model. This shouldn't be too difficult, but it will be a breaking change, especially for providers, so we should not do this for 6.0.

@stevendarby I agree that making the factory part of the key is a bit odd. My guess is that this was done to avoid capturing the factory in the lambda. Regardless, as long as the factory is in the cache it will still hold on to the scoped service, and so I don't think removing the factory from the cache key will solve this issue.

@FStolte The workaround for this is to configure the required value generator explicitly in the model, using `HasValueGeneratorFactory` or `HasValueGenerator`.

**Comment by stevendarby** [comment]
@ajcvickers I don't think that the factory _is_ in the cache. The factory is used to create the cache item. In a GetOrAdd, if the item doesn't already exist the factory is invoked and the result is cached - I don't think the factory is held on to? And of course if the item already exists the factory isn't used at all on subsequent calls.

It's this bit I'm thinking of:

```C#
return _cache.GetOrAdd(new CacheKey(property, entityType, factory), ck => ck.Factory(ck.Property, ck.EntityType));
```

If it became:

```C#
// CacheKey created without the factory; given factory is used directly rather than through the key
return _cache.GetOrAdd(new CacheKey(property, entityType), ck => factory(ck.Property, ck.EntityType));
```

wouldn't that improve the situation? The `factory` isn't added to the key and it's used once to create the cache item. I may well be missing something here, if so, sorry to waste your time!

**Comment by ajcvickers** [comment]
@stevendarby You're right. I was thinking the factory was stored in the cache, but it isn't.
